src/index.js

```diff
@@ -1,198 +1,49 @@
 'use strict'
 
-const debug = require('debug')
-const log = debug('libp2p:tcp')
-const tcp = require('net')
-const multiaddr = require('multiaddr')
-const Address6 = require('ip-address').Address6
+const net = require('net')
+const toPull = require('stream-to-pull-stream')
 const mafmt = require('mafmt')
-// const parallel = require('run-parallel')
 const contains = require('lodash.contains')
-const os = require('os')
+const isFunction = require('lodash.isfunction')
+const noop = require('lodash.noop')
 const Connection = require('interface-connection').Connection
+const debug = require('debug')
+const log = debug('libp2p:tcp')
 
-exports = module.exports = TCP
-
-const IPFS_CODE = 421
-const CLOSE_TIMEOUT = 2000
-
-function TCP () {
-  if (!(this instanceof TCP)) {
-    return new TCP()
-  }
+const createListener = require('./listener')
 
-  this.dial = function (ma, options, callback) {
-    if (typeof options === 'function') {
+module.exports = class TCP {
+  dial (ma, options, callback) {
+    if (isFunction(options)) {
       callback = options
       options = {}
     }
 
     if (!callback) {
-      callback = function noop () {}
+      callback = noop
     }
 
-    const socket = tcp.connect(ma.toOptions())
-    const conn = new Connection(socket)
-
-    socket.on('timeout', () => {
-      conn.emit('timeout')
-    })
-
-    socket.once('error', (err) => {
-      callback(err)
-    })
-
-    socket.on('connect', () => {
-      callback(null, conn)
-      conn.emit('connect')
-    })
+    const cOpts = ma.toOptions()
+    log('Connecting to %s %s', cOpts.port, cOpts.host)
+    const socket = toPull.duplex(net.connect(cOpts, callback))
 
-    conn.getObservedAddrs = (cb) => {
+    socket.getObservedAddrs = (cb) => {
```

Comment:
Can we keep the overloading/adding of methods in the 'Connection' obj. I understand that this socket is similar, but just to avoid confusion in the future by other reviewers/implementers of other transports, sockets are a thing, but libp2p Connections have more features